src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1188,491 @@ where
     }
 }
 
+#[async_trait]
+impl<Types: NodeType, State: MerklizedState<Types>> UpdateStateData<Types, State> for SqlStorage
+where
+    State::Index: Serialize + Send + Sync + Display + FromStr + 'static,
+    State::Element: Serialize + Send + Sync + 'static,
+    State::NodeValue: Send + 'static,
+{
+    async fn insert_merkle_nodes(
+        &mut self,
+        name: &'static str,
+        path: MerklePath<State::Element, State::Index, State::NodeValue>,
+        traversal_path: Vec<usize>,
+        block_number: u64,
+    ) -> QueryResult<()> {
+        let block_number = block_number as i64;
+
+        let mut traversal_path = traversal_path.iter();
+        let txn = self.transaction().await?;
+
+        // all the node hashes are collected here for batch insert at the end
+        let mut hashes = Vec::new();
+        // All the nodes are also collected here, They depend on the hash ids which are returned after hashes are upserted in the db
+        let mut nodes = Vec::new();
+
+        for node in path.iter() {
+            match node {
+                MerkleNode::Empty => {
+                    let ltree_path = LTree::from(traversal_path.clone());
+                    hashes.push([0_u8; 32].to_vec());
+                    nodes.push(Node {
+                        pos: ltree_path,
+                        ..Default::default()
+                    });
+                }
+                MerkleNode::ForgettenSubtree { .. } => {
+                    return Err(QueryError::Error {
+                        message: "Node in the Merkle path contains a forgetten subtree".to_owned(),
+                    });
+                }
+                MerkleNode::Leaf { value, pos, elem } => {
+                    let mut leaf_commit = Vec::new();
+                    // Serialize the leaf node hash value into a vector
+                    value
+                        .serialize_compressed(&mut leaf_commit)
+                        .map_err(ParseError::Serialize)?;
+
+                    // Define the LTREE path to be inserted for the leaf node.
+                    // Note:
+                    // The traversal path is from the leaf to the root.
+                    // The LTREE path is created by reversing the traversal path. Root node is represented as an empty LTree
```

Comment:
I think this comment would make more sense at the definition of `LTree::from` where this logic is actually implemented